### PR TITLE
match-description options

### PR DESCRIPTION
**Pull request by brettz9**
- BREAKING CHANGE(require-description): remove `noDefaults` option and change `contexts` to always override defaults
- feat(match-description): allow `mainDescription: string|boolean` to override or disable main description separate from default
- feat(match-description): add `contexts` options for control on which contexts the rules apply
- feat(match-description): report line number with `match-description`
- feat(match-description): allow reporting multiple errors when main description validation fails
- feat(require-jsdoc): add `contexts` option to allow working with any node type
- fix(match-description, require-description): allow `contexts` to work with any node type
- fix(match-description): tighten default regex to require punctuation at the end even if only a single character

### CODE REVIEW

File: test/rules/assertions/matchDescription.js [hunk 1]
```diff
@@ -41,11 +43,34 @@ export default {
...
+      `,
+      errors: [
+        {
+          line: 3,
+          message: 'JSDoc description does not satisfy the regex pattern.'
+        }
+      ],
+      options: [{
+        tags: {
+          'main description': '[\u0410-\u042F][\u0410-\u044F]+\\.',
```
Comment (golopot) [inline test/rules/assertions/matchDescription.js]: 1. Shouldn't it be in camelcase? 
2. Imo it is better to place the `mainDescription` in the top level of the option. Since a) `mainDescription` is not a tag name, b) Putting mainDescription in top-level allows schema validation to prevent typo.

File: src/rules/matchDescription.js [hunk 2]
```diff
@@ -47,18 +52,36 @@ export default iterateJsdoc(({
...
+            oneOf: [
+              {
+                items: {
+                  type: 'string'
+                },
+                type: 'array'
+              },
+              {
+                type: 'string'
+              }
```
Comment (golopot) [inline src/rules/matchDescription.js]: I prefer allowing only `array` to keep it simple.

File: src/rules/matchDescription.js [hunk 3]
```diff
@@ -47,18 +52,36 @@ export default iterateJsdoc(({
...
+            ]
+          },
           matchDescription: {
             format: 'regex',
             type: 'string'
           },
+          noDefaults: {
```
Comment (golopot) [inline src/rules/matchDescription.js]: Why this option? Why not just setting `contexts`?

File: .README/rules/require-jsdoc.md [hunk 4]
```diff
@@ -34,6 +34,11 @@ be checked by the rule.
   - `FunctionExpression`
   - `MethodDefinition`
 
+- `contexts` - Set this to a string or array of strings representing the additional
```
Comment (golopot) [inline .README/rules/require-jsdoc.md]: This belongs to another pr.

File: test/rules/assertions/matchDescription.js (hunk 1 above)
Comment (brettz9) [reply test/rules/assertions/matchDescription.js]: I chose "main description" as a special case as users couldn't have a tag with this name as it has a space in it, but I see your point re: allowing schema validation (even though it adds a little extra complexity to the code).

File: src/rules/matchDescription.js (hunk 2 above)
Comment (brettz9) [reply src/rules/matchDescription.js]: I prefer accepting string for simplicity to the user. :) @gajus ?

File: src/rules/matchDescription.js (hunk 3 above)
Comment (brettz9) [reply src/rules/matchDescription.js]: It is extra work for users. I believe people will almost always want to keep the defaults. Also, that is the existing behavior for `require-description`.

File: .README/rules/require-jsdoc.md (hunk 4 above)
Comment (brettz9) [reply .README/rules/require-jsdoc.md]: I can move it to another PR, but note that it all leverages the same code.

**Comment by brettz9** [comment]
`contexts` is very useful as one can, as per the request in #293 , ensure that one has valid descriptions, e.g., above properties--not only the default behavior of matching descriptions on functions only.

**Comment by golopot** [comment]
I mean the rule should apply to jsdoc attached to all node types and there is no configuration to include or exclude.

**Comment by brettz9** [comment]
I think it is better to be under user control, whether for performance or just because projects differ from one another in their needs; but if you want to expand the defaults, I can change to support all node types by default.

File: test/rules/assertions/matchDescription.js (hunk 1 above)
Comment (brettz9) [reply test/rules/assertions/matchDescription.js]: Done, thanks!

File: src/rules/matchDescription.js (hunk 2 above)
Comment (gajus) [reply src/rules/matchDescription.js]: Please only accept `array`. Multiple APIs create ambiguity about how to supply multiple parameters.

File: .README/rules/match-description.md [hunk 5]
```diff
@@ -41,14 +50,39 @@ tag should be linted with the `matchDescription` value (or the default).
...
+}
+```
+
+There is no need to add `"main description": true`, as by default, the main
+function (and only the main function) is linted, though you may disable checking
+it by setting it to `false`.
+
+##### `contexts`
+
+Set this to a string or array of strings representing the AST context
```
Comment (gajus) [inline .README/rules/match-description.md]: Just array of strings.

File: src/rules/matchDescription.js (hunk 3 above)
Comment (gajus) [reply src/rules/matchDescription.js]: > Why this option? Isn't it more intuitive to let provided contexts to override the default context?

Agreed. Otherwise, I need to be aware of what the defaults are. `contexts` should just override the default.

Comment (brettz9) [reply src/rules/matchDescription.js]: There have been a few rules already added using this approach, e.g., `require-description`, `check-types`. Revert them as well?

Comment (brettz9) [reply src/rules/matchDescription.js]: (`check-types` is somewhat different in referring to avoid the defaults for your preferred types, while `require-description` is more like as in this PR.)

Comment (brettz9) [reply src/rules/matchDescription.js]: I went ahead and added a commit to make a breaking change to remove `noDefaults` from `require-description` (and for `match-description`) though I haven't done so for `check-types` unless you want to confirm. Let me know if that is ok.

File: src/rules/matchDescription.js (hunk 2 above)
Comment (brettz9) [reply src/rules/matchDescription.js]: Ok, done.

File: .README/rules/match-description.md (hunk 5 above)
Comment (brettz9) [reply .README/rules/match-description.md]: Ok, sure, done. Also for `require-jsdoc` and `require-description`.

File: src/rules/matchDescription.js (hunk 3 above)
Comment (brettz9) [reply src/rules/matchDescription.js]: @gajus : Since your approval kind of overlapped with my reversion, I just wanted to confirm--are you happy with my breaking change to undo `require-description`'s use of `noDefaults` (in addition to removing it in `match-description`)?

**Comment by golopot** [comment]
In the future please keep unrelated changes in different pull requests. Stuffing all changes in a pr 1) makes it much harder to review, 2) make the reviewer unable to accept one part of change and reject another.

**Comment by brettz9** [comment]
This PR has now been narrowed somewhat, and I think it is ready for merging, @gajus , if you could respond to my comments ending at https://github.com/gajus/eslint-plugin-jsdoc/pull/294#discussion_r297063836 mentioning how I added a breaking change that I understood you wanted around the time you approved this PR, so I wanted to confirm that breaking change was indeed ok.

**Comment by brettz9** [comment]
If y'all are ready for rebasing, sure.

File: src/rules/requireJsdoc.js [hunk 6]
```diff
@@ -7,6 +7,12 @@ import getJSDocComment from '../eslint/getJSDocComment';
 const OPTIONS_SCHEMA = {
   additionalProperties: false,
   properties: {
+    contexts: {
```
Comment (golopot) [inline src/rules/requireJsdoc.js]: This option overlapped the option `require`?

Comment (brettz9) [reply src/rules/requireJsdoc.js]: Good catch. Yes, I avoided a breaking change here in order to allow `require-jsdoc` to offer more arbitrary contexts while also phasing in the API used by `require-description` and `match-description`. I can make a breaking change to remove `require` in favor of `contexts` if you and @gajus wish. I think it is a good idea.

Comment (golopot) [reply src/rules/requireJsdoc.js]: I am ok either way.

Comment (gajus) [reply src/rules/requireJsdoc.js]: Regarding breaking-changes: I am Okay with as many breaking changes as you need. Far bigger preference for clean, unambiguous APIs without legacy remnants over how often a breaking-change is made. This seems like an unpopular opinion in JavaScript land, but in my opinion a greater number of breaking changes over time is better (as it allows more gradual adoption, one version at a time) than one major planned breaking change that batches together many unrelated API changes and requires significant time investment to upgrade.

Comment (brettz9) [reply src/rules/requireJsdoc.js]: Yes, it is jarring at first when one sees a "major" change and expects it will require a huge number of changes, but if people get to know the repository/package (or note how large its major version already has become), and how it may do frequent though incremental breaking releases, the angst, as well as the incremental effort is less.
